File: examples/basic_usage.py
"""
Example: Basic usage of GeminiSDK.

This example demonstrates:
- Creating a GeminiClient
- Creating a session
- Sending messages with streaming
- Using event handlers
"""

import asyncio

from geminisdk import GeminiClient, EventType


async def basic_example():
    """Basic example showing streaming chat."""
    print("=== Basic Streaming Example ===\n")
    
    async with GeminiClient() as client:
        # Create a session
        session = await client.create_session({
            "model": "gemini-2.5-pro",
            "streaming": True,
        })
        
        # Subscribe to events
        def on_event(event):
            if event.type == EventType.ASSISTANT_MESSAGE_DELTA:
                # Print streaming content without newline
                print(event.data["delta_content"], end="", flush=True)
            elif event.type == EventType.ASSISTANT_MESSAGE:
                # Final message received
                print("\n\n--- Response complete ---")
            elif event.type == EventType.SESSION_ERROR:
                print(f"\nError: {event.data['error']}")
        
        session.on(on_event)
        
        # Send a message
        print("User: What is Python?\n")
        print("Assistant: ", end="")
        
        await session.send({
            "prompt": "What is Python? Explain in 2-3 sentences.",
        })


async def non_streaming_example():
    """Example using non-streaming send_and_wait."""
    print("\n\n=== Non-Streaming Example ===\n")
    
    async with GeminiClient() as client:
        session = await client.create_session({
            "model": "gemini-2.5-flash",
            "streaming": False,
        })
        
        print("User: Write a haiku about coding.\n")
        
        response = await session.send_and_wait({
            "prompt": "Write a haiku about coding.",
        })
        
        print(f"Assistant: {response.data['content']}")


async def main():
    """Run examples."""
    try:
        await basic_example()
        await non_streaming_example()
    except Exception as e:
        print(f"\nError: {e}")
        print("\nMake sure you've authenticated with Gemini CLI:")
        print("  npm install -g @google/gemini-cli")
        print("  gemini auth login")


if __name__ == "__main__":
    asyncio.run(main())

File: src/cpp/examples/basic_usage.cpp
/**
 * @file basic_usage.cpp
 * @brief Basic usage example for GeminiSDK C++
 */

#include <geminisdk/geminisdk.hpp>
#include <iostream>

using namespace geminisdk;

int main() {
    std::cout << "GeminiSDK C++ - Basic Usage Example\n\n";
    
    try {
        // Create client with default options
        Client client;
        
        // Start the client (authenticates with Gemini CLI credentials)
        std::cout << "Starting client...\n";
        client.start();
        std::cout << "Client started and authenticated!\n\n";
        
        // List available models
        std::cout << "Available models:\n";
        for (const auto& model : client.list_models()) {
            std::cout << "  - " << model.name << " (" << model.id << ")\n";
        }
        std::cout << "\n";
        
        // Create a session
        std::cout << "Creating session...\n";
        SessionConfig config;
        config.model = "gemini-2.5-flash";
        config.streaming = false;  // Non-streaming for simplicity
        config.system_message = "You are a helpful assistant.";
        
        auto session = client.create_session(config);
        std::cout << "Session created: " << session->session_id() << "\n\n";
        
        // Send a message and wait for response
        std::cout << "Sending message...\n";
        MessageOptions options;
        options.prompt = "What are three interesting facts about the C++ programming language?";
        
        auto response = session->send_and_wait(options);
        
        std::cout << "Response received:\n";
        std::cout << "Event type: " << event_type_to_string(response.event_type) << "\n";
        if (response.data.contains("content")) {
            std::cout << "Content: " << response.data["content"].get<std::string>() << "\n";
        }
        
        // Clean up
        std::cout << "\nClosing client...\n";
        client.close();
        std::cout << "Done!\n";
        
    } catch (const GeminiSDKError& e) {
        std::cerr << "GeminiSDK Error: " << e.what() << "\n";
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    
    return 0;
}
